core/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java

```diff
@@ -60,8 +60,10 @@
     public static final ParseField TYPE_FIELD = new ParseField("type");
     public static final ParseField QUERY_FIELD = new ParseField("query");
 
-    /** The default name for the match query */
-    public static final String NAME = "match";
+    /** The names for the match query. */
+    // TODO replace this with a Types like object
+    public static final Names NAMES = new Names("match", "match_phrase", "matchPhrase", "match_phrase_prefix", "matchPhrasePrefix",
```

Comment:
I've just noticed in the parsing code that option 1 above is already present (you can specify the `type` in the body of the query). I think given this we should definitely deprecate the `match_phrase` and `match_phrase_prefix` names since you can at the moment do something wholly confusing like:

```
{
  "query": {
    "match_phrase": {
      "my_field": {
        "query": "foo",
        "type": "boolean"
      }
    }
  }
}
```

The above query looks like a `match_phrase` query but will actually override the `type` and run as a `boolean` match query.